pkg/backup/file_sst_sink.go

```diff
@@ -331,11 +349,217 @@ func (s *fileSSTSink) write(ctx context.Context, resp exportedSpan) (roachpb.Key
 	return resp.resumeKey, err
 }
 
-// adjustFileEndKey checks if the export respsonse end key can be used as a
-// split point during restore. If the end key is not splitable (i.e. it splits
+// writeKey writes a single key to the SST file. The key should be the full key. span refers to the span that the key
+// belongs to. start and end are the time bounds of the span being backed up. The writing of each key could
+// potentially split up the span of the previously written key (if both keys were part of the same span and the span
+// had to be split). As a consequence the span that is recorded for the new key is returned, as it may be a subspan of
+// the span passed in for the key.
+//
+// flush should be called after the last key is written to ensure that the SST is written to the destination.
```

Comment:
One of the biggest hurdles for me writing this was determining some of the implicit assumptions we make about the data, which wasn't laid out clearly in `write`.

I think this could maybe use some additional elaborations? The behavior I have laid out mentally is as so:

`writeKey` in its most simplest form will write a key such that the previous key was part of the same span, in which case it just adds to that previous `BackupManifest_File`. In the event that the previous file has reached the `fileSpanByteLimit`, the previous file will have its span exclusive end updated to this new key, and then create a new file with the new key as its span start, and the end is left untouched.

Another scenario is `writeKey` receives a `span` that contiguously extends the last `BackupManifest_File`'s span (i.e. `span.Key == lastFile.Span.EndKey`). In that case, we can extend the previous backup manifest file by updating its end key to the new span's end key and continue writing. In the event we are unable to do this because of size constraints, we don't  do any extension and just treat the `span` as a new file.

These are the main cases and what I expect to encounter in a compaction. I did add some guardrails against edge cases, namely where the span precedes the previous `BackupManifest_File` or if the key being written precedes the last written key. The behavior here is to just flush. From a backup perspective, this could potentially result in `BackupManifest_File`s overlapping within the same backup. However, I think that's fine as this still creates valid `BackupManifest_File`s with the correct backing SSTs.